### DISCUSSION TITLE
Can I use DrawImage on the same image from different threads?

### DISCUSSION THREAD

**Discussion by Raildex** [SixLabors/ImageSharp#2729] [Q&A]
I generate "Buckets" which represent a small portion of a bigger image. The bucket images have their own memory.
Each bucket belongs to a thread.
Each Thread then uses
```cs
image.Mutate(l => l.DrawImage(bucketImage,new Point(bucket.X,bucket.Y),new Rectangle(0,0,bucket.W,bucket.H), PixelColorBlendingMode.Add, 1.0f));
```

The bucket dimensions (Bucket.X,Bucket.Y,Bucket.W,Bucket.H) do not overlap.

**Comment by JimBobSquarePants** [comment]
ImageSharp already stores larger buffers using discontiguous chunks so it's likely already split up. 

It should be thread safe though so your approach should work.

**Reply by antonfirsov** [reply]
Do we *guarantee* `Image` APIs to be thread-safe though?

I would advice against this for performance reasons. Depending on the bucket sizes, and the size of `image`, this it might end up being very inefficient because of [false sharing](https://en.wikipedia.org/wiki/False_sharing).

I would rather implement this with producer/consumer pattern, where the bucket threads post work to a thread owning the big `image`. Maybe [System.Threading.Channels](https://learn.microsoft.com/en-us/dotnet/core/extensions/channels) can help implementing this?

**Reply by JimBobSquarePants** [reply]
There’s no written guarantee but the implementations should be. Brushes were an issue in the past but we fixed that together. 

Good link re false sharing!

**Comment by Raildex** [comment]
I've come back to Imagesharp and noticed that `DrawImage` isn't threadsafe.
My code is fairly complex and I will try to make an example, but Imagesharp does not seem to be thread safe when concurrently writing to an image.

Just as in my OP, i have a big image and use many smaller images (using Image.Wrap onto a RgbaVector Memory) to draw onto the bigger image.

The result when using C#'s `lock` on the big image:

![Room 0vzcinz5a zgd_0](https://github.com/user-attachments/assets/4042ad72-5785-4efb-8d87-10dece9abb15)


the Result when not locking the big image:

![Room 0sr0bqkcb 540_1](https://github.com/user-attachments/assets/ae5e955e-411d-42c6-afdf-12a55b7fba9b)

Of course this is highly dependent on what I do and how many threads I use and like all race conditions this is not easy to reproduce. however, I get fairly consistent "artifacts" when removing the lock.

**Comment by JimBobSquarePants** [comment]
You’ll need to share some code here so I can replicate.
